Add docs about CAPI node pools (#1991)
giantswarm/docs: src/content/advanced/cluster-management/node-pools-capi/index.md

```yaml
metadata:
  description: "my cluster"
  name: test-cluster
  organization: giantswarm
nodePools:
  pool0:
    maxSize: 2
    minSize: 2
    spotInstances:
      enabled: true
      maxPrice: 1.2
```

{{< /tab >}}
{{< tab id="cluster-capz-azure-vms" for-impl="capz_vms" >}}

This is currently not supported.

{{< /tab >}}
{{< /tabs >}}

## Limitations

- Clusters without worker nodes (= without node pools) cannot be considered fully functional. In order to have all
required components scheduled, worker nodes are required. For that reason, we deactivate any monitoring and alerts for
these clusters and don't provide any proactive support.

### Things to keep in mind

{{< tabs >}}
{{< tab id="cluster-capa-ec2" for-impl="capa_ec2" >}}

- At times, the EC2 instance type required by a node pool can be unavailable in certain availability zones. This can
result in node pools providing less than the desired number of nodes. The more availability zones a node pool spans,
the less likely this problem is to occur.

> - By default, clusters can have up to 5 node pools. This is limited by the AWS service quota named "IPv4 CIDR blocks

Reviewer: this is only true for vintage

Reviewer: In CAPI we do not add CIDR block for new node pool so the limit do not apply

---
linkTitle: Node pools (CAPI)
title: Node pools (CAPI)
description: A general description of node pools as a concept, its benefits, and some details you should be aware of.

> weight: 50

Reviewer 2: ```suggestion
weight: 45
```

to be before vintage for sure (not sure what Hugo does for equal weight).

menu:
  main:
    parent: advanced-cluster-management
user_questions:
  - What is a node pool?
  - What are node pools?
  - In which cloud environments are node pools supported?
  - Which workload cluster releases introduced node pools?
owner:
  - https://github.com/orgs/giantswarm/teams/team-phoenix
last_review_date: 2023-11-22
---

## Definition

> A node pool is a set of nodes within a Kubernetes cluster that share the same configuration (machine type, CIDR range, etc.). Each node in the pool is labeled by the node pool's name.

Reviewer: ` CIDR range` I would remove this as all nodes in whole cluster share same CIDR range,

Different CIDR range per nodepool was true for vintage

## Advantages

Prior to the introduction of node pools, a cluster could only comprise one type of worker node. The cluster would have
to be scaled as a whole, and the availability zone distribution would apply to all worker nodes of a cluster. This
would mean that every worker node would have to be big enough to run the largest possible workload, in terms of
memory and CPU resources required. At the same time, all worker nodes in the cluster would have to use the same
availability zone distribution, even if some workloads wouldn't require the increased availability.

Node pools are independent groups of worker nodes belonging to a cluster, where all nodes within a pool share a
common configuration. You can combine any type of node pool within one cluster. Node pools can differ regarding:

- Machine type
- Availability zone distribution

> - Scaling configuration (number of nodes)

Reviewer 2: ```suggestion
- Scaling configuration (number of nodes)
- Node labels (the pool name is added as node label by default)
```

and then below, we explain that these can be used to run workloads on specific node pools (high RAM, GPU support, specific AZ, etc.)

> A node pool is identified by a name that you can pick as a cluster administrator. The name of the node pool is set as a label on all nodes belonging to the pool.

Reviewer: This is a good place to mention limitation of the nodepool name, like how many chars and waht chars are allowed + few examples

Reviewer 2: And let's also be specific. The introduction already talked about name labeling. Let's reference the actual `giantswarm.io/machine-pool=...` label here. Or since you already specify it below in the per-provider section, maybe rather remove this sentence here (see also suggestion above to add it as bullet point).

## Configuration

> Node pools can be created, deleted or updated by changing the configuration used when creating the cluster using [`kubectl`]({{< relref "/getting-started/create-workload-cluster" >}})

Reviewer: `kubectl` -> `kubectl gs`

{{< tabs >}}
{{< tab id="cluster-capa-ec2" for-impl="capa_ec2" >}}

```nohighlight
kubectl gs template cluster --provider capa --name a1b2c3 \
  --organization giantswarm \
  --description "my test cluster" \
  --machine-pool-name pool0 \
  --machine-pool-min-size 3 \
  --machine-pool-max-size 5 \
  --machine-pool-instance-type r6i.xlarge
```

The node pool name will be prepended by the cluster name. In the example above, the node pool name will be `a1b2c3-pool0`.

> All nodes in the node pool will be labeled with the node pool name, using the  `giantswarm.io/machine-pool` label.

Reviewer 2: ```suggestion
All nodes in the node pool will be labeled with the node pool name, using the `giantswarm.io/machine-pool` label.
```

You can identify the nodes' node pool using that label.
The example `kubectl` command below will list all nodes with role, node pool name, and node name.

```nohighlight
kubectl get nodes \
  -o=jsonpath='{range .items[*]}{.metadata.labels.kubernetes\.io/role}{"\t"}{.metadata.labels.giantswarm\.io/machine-pool}{"\t"}{.metadata.name}{"\n"}{end}' | sort
master         ip-10-1-5-55.eu-central-1.compute.internal
worker  a1b2c3-pool0  ip-10-1-6-225.eu-central-1.compute.internal
worker  a1b2c3-pool0  ip-10-1-6-67.eu-central-1.compute.internal
```

{{< /tab >}}
{{< tab id="cluster-capz-azure-vms" for-impl="capz_vms" >}}

```nohighlight
kubectl gs template cluster --provider capz --name test-cluster \
  --organization giantswarm \
  --description "my test cluster" \
  --machine-pool-name pool0 \
  --machine-pool-min-size 3 \
  --machine-pool-max-size 5 \

>   --machine-pool-instance-type r6i.xlarge

Reviewer 2: This looks like an AWS instance type, not Azure VM type.

```

The node pool name will be prepended by the cluster name. In the example above, the node pool name will be `a1b2c3-pool0`.

> All nodes in the node pool will be labeled with the node pool name, using the  `giantswarm.io/machine-deployment` label.

Reviewer 2: ```suggestion
All nodes in the node pool will be labeled with the node pool name, using the `giantswarm.io/machine-deployment` label.
```

You can identify the nodes' node pool using that label.
The example `kubectl` command below will list all nodes with role, node pool name, and node name.

```nohighlight
kubectl get nodes \
  -o=jsonpath='{range .items[*]}{.metadata.labels.kubernetes\.io/role}{"\t"}{.metadata.labels.giantswarm\.io/machine-deployment}{"\t"}{.metadata.name}{"\n"}{end}' | sort
master         ip-10-1-5-55.eu-central-1.compute.internal
worker  a1b2c3-pool0  ip-10-1-6-225.eu-central-1.compute.internal

> worker  a1b2c3-pool0  ip-10-1-6-67.eu-central-1.compute.internal

Reviewer 2: Same as above: these are AWS instance name examples, not Azure VM names

```

{{< /tab >}}
{{< /tabs >}}

## Assigning workloads to node pools {#assigning-workloads}

> Knowing the node pool name of the pool to use, you can use the `nodeSelector` method of assigning pods to the node pool.

Reviewer 2: ```suggestion
Knowing the node pool name of the pool to use, you can use the [`nodeSelector` method of assigning pods](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#nodeselector) to the node pool.
```

> Assuming that the node pool name is `pool0`, and the cluster name is `a1b2c3`, your `nodeSelector` could for example look like this:

Reviewer 2: I suggest `s/a1b2c3/mycluster/` everywhere, since that name is used in https://docs.giantswarm.io/getting-started/create-workload-cluster/#step-2-template-and-create-the-workload-cluster. Some consistency would be great, and those random names are harder to read.

{{< tabs >}}
{{< tab id="cluster-capa-ec2" for-impl="capa_ec2" >}}

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: nginx
spec:
  containers:
  - name: nginx
    image: nginx
  nodeSelector:
    giantswarm.io/machine-pool: a1b2c3-pool0
```

{{< /tab >}}
{{< tab id="cluster-capz-azure-vms" for-impl="capz_vms" >}}

```yaml
A similar example for an Azure cluster:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: nginx
spec:
  containers:
  - name: nginx
    image: nginx
  nodeSelector:
    giantswarm.io/machine-deployment: a1b2c3-pool0
```

{{< /tab >}}
{{< /tabs >}}

You can assign workloads to node pools in a more indirect way too. This is achieved by using other node attributes which are

> specified via the node pool and which are exposed as node labels.

Reviewer 2: Can we rephrase this so it refers to automatically generated labels? It currently reads as if one needs to explicitly specify them. The given examples such as `topology.kubernetes.io/zone` are auto-generated. We don't need to explain in detail which controller does that, but maybe we can simply refer to well-known labels as listed at https://kubernetes.io/docs/reference/labels-annotations-taints/. Together with the below examples, that's enough.

> For example: In a case where you have node pools with one instance type. Using a `nodeSelector` with the label `node.kubernetes.io/instance-type` you can assign workloads to matching nodes only.
> 
> Another example: In a case where you have different node pools using different availability zones. With a `nodeSelector` using the label `topology.kubernetes.io/zone` you can assign your workload to the nodes in a particular availability zone.

Reviewer 2: ```suggestion
For example: you have several node pools with different instance types. Using a `nodeSelector` with the label `node.kubernetes.io/instance-type`, you can assign workloads only to matching nodes.

Another example: you have different node pools using different availability zones. With a `nodeSelector` using the label `topology.kubernetes.io/zone`, you can assign your workload to the nodes in a particular availability zone.
```

## Adding more node pools

You can add new node pools at any time. You just need to update the cluster configuration and specify the details of the node pools that you want to add.

For example, if this was the cluster configuration

```yaml
metadata:
  description: "my cluster"
  name: test-cluster
  organization: giantswarm
nodePools:
  nodepool0:
    maxSize: 4
    minSize: 3
    instanceTypeOverrides:
    - r6i.xlarge
    - r5.xlarge

>     - m5.xlarge

Reviewer 2: `nodepool0` should remain the same in the before/after YAML

```

We can add a new node pool like this

```yaml
metadata:
  description: "my cluster"
  name: test-cluster
  organization: giantswarm
nodePools:
  nodepool0:
    instanceType: r6i.xlarge
    maxSize: 4
    minSize: 3
  nodepool1:
    instanceType: m5.xlarge
    maxSize: 3
    minSize: 1
```

> All the nodes from each node pool will be labeled accordingly.
> 
> ## Updating an existing node pool

Reviewer 2: ```suggestion
## Updating an existing node pool
```

Not necessary to repeat this, as it was clearly explained above.

Instances in the node pool will be rolled whenever these properties are changed in the node pool definition:
- `instanceType`
- `additionalSecurityGroups`

Instances will also be rolled if these values are changed:
- `providerSpecific.ami`

### What happens when a node pool is updated

When node pool instances need to be rolled, each instance receives a terminate signal from AWS.
This is propagated as shutdown signal to the OS and then to each running process like the `kubelet`, which will send a `NodeShutDown` event to the pod.

The `kubelet` will wait up to 5 minutes for all pods to terminate and after that it will terminate itself and the shutdown of the AWS EC2 instance will finally proceed.

> Please, be aware that it may happen that AWS decides to force terminate the instance before the 5 minutes.

Reviewer 2: ```suggestion
Please, be aware that it may happen that AWS decides to force-terminate the instance before the 5 minutes. We recommend draining the respective nodes and moving workloads to other nodes before such an update operation.
```

## Node pool deletion

> TBD

Reviewer 2: Until this is solved, maybe we can add a tab for vintage already if we support deletion there (I guess since it's a separate CR without Helm "keep" annotation).

Author: But we have a different page for `vintage`. The idea is to document vintage on its own page, and use this one only for CAPI.

> ## Using mixed instance types {#mixed-instance-types}

Reviewer 2: ```suggestion
## Using mixed instance types (only {{% impl_title "capa_ec2" %}}) {#mixed-instance-types}
```

**Note:** This feature was called ["Similar instance types" on vintage]({{< relref "/advanced/cluster-management/node-pools-vintage#similar-instance-types" >}}).

> On AWS EC2 (CAPA) you can override the instance type of the node pool to enable [mixed instances policy on AWS](https://docs.aws.amazon.com/autoscaling/ec2/APIReference/API_LaunchTemplateOverrides.html).

Reviewer 2: ```suggestion
On {{% impl_title "capa_ec2" %}}, you can override the instance type of the node pool to enable [mixed instances policy on AWS](https://docs.aws.amazon.com/autoscaling/ec2/APIReference/API_LaunchTemplateOverrides.html).
```

Also, the order in which we define the instance types in `instanceTypeOverrides` is important.
The first instance type in the list that is available in the selected availability zone will be used.

**Note:** Changing the `instanceTypeOverrides` value won't trigger a rolling update of the node pool.

Using multiple instance types in a node pool has some benefits:

- Together with spot instances, using multiple instance types allows better price optimization. Popular instance types tend to have more price adjustments. Allowing older-generation instance types that are less popular tends to result in lower costs and fewer interruptions.

- Even without spot instances, AWS has a limited number of instances per type in each Availability Zone. It can happen that your selected instance type is temporarily out of stock just in the moment you are in demand of more worker nodes. Allowing the node pool to use multiple instance types reduces this risk and increases the likelihood that your node pool can grow when in need.

## Node pools and autoscaling {#autoscaling}

> With node pools, you set the autoscaling range per node pool (suppported on AWS EC2 (CAPA) clusters only). The Kubernetes cluster autoscaler has to decide which node pool to scale under which circumstances.

Reviewer 2: ```suggestion
With node pools, you set the autoscaling range per node pool (suppported on {{% impl_title "capa_ec2" %}} clusters only). The Kubernetes cluster autoscaler has to decide which node pool to scale under which circumstances.
```

If you assign workloads to node pools as described [above](#assigning-workloads) and the autoscaler finds pods in `Pending` state, it will decide based on the node selectors which node pools to scale up.

In case there are workloads not assigned to any node pools, the autoscaler may pick any node pool for scaling. For details on the decision logic, please check the upstream FAQ for [AWS](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/aws/README.md).

## On-demand and spot instances {#on-demand-spot}

{{< tabs >}}
{{< tab id="cluster-capa-ec2" for-impl="capa_ec2" >}}

> Node pools can make use of [Amazon EC2 Spot Instances](https://aws.amazon.com/ec2/spot/). You just need to enable it and select the max price to pay on the node pool definition.

Reviewer 2: ```suggestion
Node pools can make use of [Amazon EC2 Spot Instances](https://aws.amazon.com/ec2/spot/). On the node pool definition, you can enable it and select the maximum price to pay.
```

Avoiding the word "just" because it can also relay a passive-aggressive "Hey it's so easy, a baby could do it" meaning.

```yaml
metadata:
  description: "my cluster"
  name: test-cluster
  organization: giantswarm
nodePools:
  pool0:
    maxSize: 2
    minSize: 2
    spotInstances:
      enabled: true
      maxPrice: 1.2
```

{{< /tab >}}
{{< tab id="cluster-capz-azure-vms" for-impl="capz_vms" >}}

This is currently not supported.

{{< /tab >}}
{{< /tabs >}}

## Limitations

> - Clusters without worker nodes (= without node pools) cannot be considered fully functional. In order to have all
> required components scheduled, worker nodes are required. For that reason, we deactivate any monitoring and alerts for
> these clusters and don't provide any proactive support.
> 
> ### Things to keep in mind
> 
> {{< tabs >}}
> {{< tab id="cluster-capa-ec2" for-impl="capa_ec2" >}}

Reviewer 2: ```suggestion
{{< tabs >}}
{{< tab id="general" title="General" >}}

- Clusters without worker nodes (= without node pools) cannot be considered fully functional. In order to have all
required components scheduled, worker nodes are required. For that reason, we deactivate any monitoring and alerts for
clusters without worker nodes and don't provide any proactive support.

{{< /tab >}}
{{< tab id="cluster-capa-ec2" for-impl="capa_ec2" >}}
```

Reviewer 2: and please ensure that all `tab id="..."` are unique (= don't repeat `id="cluster-capa-ec2"` but use `limitations-capa-ec2`) so that the tab-switching logic works i.e. switches all equally-titled tabs at once on the page.

- At times, the EC2 instance type required by a node pool can be unavailable in certain availability zones. This can
result in node pools providing less than the desired number of nodes. The more availability zones a node pool spans,
the less likely this problem is to occur.

- By default, clusters can have up to 5 node pools. This is limited by the AWS service quota named "IPv4 CIDR blocks
per VPC" in the [VPC section](https://docs.aws.amazon.com/vpc/latest/userguide/amazon-vpc-limits.html). As the AWS
account owner, you can request an increase of this limit via the AWS console.

> - Node pools can span a maximum of four availability zones. This limit affects both the number of availability zones

Reviewer: TBH I am not sure this is true as well, I think we are limited by number of AZs in the region and also there is setting how many AZs should be used for subnets  https://github.com/giantswarm/cluster-aws/blob/master/helm/cluster-aws/templates/_aws_cluster.tpl#L56

Reviewer 2: Probably this limitation, if it's even applicable, is too detailed and most customers won't hit it. Maybe we want to remove this hint altogether, or keep only the vintage specifics on a different tab.

a single node pool can cover, as well as the number of different availability zones all node pools of a cluster can
cover. Once an availability zone has been assigned for use in a cluster, either for a control plane node or for worker
nodes, it cannot be unassigned from that cluster. It will remain assigned even if there are no more node pools using
that availability zone.

>     - **Example:** The control plane node(s) are in availability zone A. Node pool 1 uses availability zones B and C. Node pool 2 uses

Reviewer: This is only true for vintage
